lib/dm-migrations/adapters/dm-do-adapter.rb

```diff
@@ -197,8 +197,14 @@ def create_unique_index_statements(model)
         def property_schema_hash(property)
           dump_class = property.dump_class
           type_map   = self.class.type_map
+          next_property_class = property.class
+          type_by_property_class = nil
+          while (type_by_property_class.nil? && next_property_class < DataMapper::Property) do
```

Comment:
You could get the property.class + its ancestors, select the ones that are `Property` descendants, map that into the type_map, compact, and select the last.